clients/include/utility.h

```diff
@@ -525,63 +530,97 @@ class Arguments
+struct RocBLAS_Data
+{
+    // filter iterator
+    typedef boost::filter_iterator<std::function<bool(const Arguments&)>,
+                                   istream_iterator<Arguments>>
+        iterator;
+
+    // Initialize class
+    static void init(const string& file) { get(file); }
+
+    // begin() iterator which accepts an optional filter.
+    static iterator begin(std::function<bool(const Arguments&)> filter = [](const Arguments&) {
+        return true;
+    })
     {
-        M = rhs.M;
-        N = rhs.N;
-        K = rhs.K;
-
-        lda = rhs.lda;
-        ldb = rhs.ldb;
-        ldc = rhs.ldc;
-        ldd = rhs.ldd;
-
-        a_type       = rhs.a_type;
-        b_type       = rhs.b_type;
-        c_type       = rhs.c_type;
-        d_type       = rhs.d_type;
-        compute_type = rhs.compute_type;
-
-        incx = rhs.incx;
-        incy = rhs.incy;
-        incd = rhs.incd;
-        incb = rhs.incb;
-
-        start = rhs.start;
-        end   = rhs.end;
-        step  = rhs.step;
-
-        alpha = rhs.alpha;
-        beta  = rhs.beta;
-
-        transA_option = rhs.transA_option;
-        transB_option = rhs.transB_option;
-        side_option   = rhs.side_option;
-        uplo_option   = rhs.uplo_option;
-        diag_option   = rhs.diag_option;
-
-        apiCallCount = rhs.apiCallCount;
-        batch_count  = rhs.batch_count;
-
-        norm_check = rhs.norm_check;
-        unit_check = rhs.unit_check;
-        timing     = rhs.timing;
-
-        algo           = rhs.algo;
-        solution_index = rhs.solution_index;
-        flags          = rhs.flags;
-        workspace_size = rhs.workspace_size;
-
-        return *this;
+        auto& ifs = get().ifs;
+
+        // We re-seek the file back to position 0
+        ifs.clear();
+        ifs.seekg(0);
+
+        // We create a filter iterator which will choose only those test cases
+        // we want right now. This is to preserve Gtest output structure while
+        // not creating no-op tests which "always pass".
+        return iterator(filter, istream_iterator<Arguments>(ifs));
     }
+
+    // end() iterator
+    static iterator end() { return iterator(); }
+
+    private:
+    // We define this function to generate a single instance of the class on
+    // first use so that we don't depend on the static initialization order.
+    // Only the first call needs to specify file.
+    static RocBLAS_Data& get(const string& file = "<unspecified file>")
+    {
+        static RocBLAS_Data singleton(file);
+        return singleton;
+    }
+
+    // Constructor which opens file
+    explicit RocBLAS_Data(const string& file)
+    {
+        ifs.open(file, ifstream::binary);
+        if(ifs.fail())
+        {
+            cerr << "Cannot open " << file << ": " << strerror(errno) << endl;
+            throw ifstream::failure("Cannot open " + file);
+        }
+    }
+
+    ifstream ifs;
 };
 
+typedef RocBLAS_Data<rocblas_test_data> RocBLAS_TestData;
```

Comment:
Since the RocBLAS_Data class template is static and only consists of single instances, we have to instantiate it once for the correctness tests and once for the performance tests.